### ISSUE TITLE
mount rootfs fails unless rootdelay=1 added

### ISSUE THREAD

**Issue by garycramblitt** [linux-sunxi/linux-sunxi#32]
Building 4GB SD card following my own instructions at http://rhombus-tech.net/allwinner_a10/hacking_the_mele_a1000/Building_Debian_From_Source_Code_for_Mele/. Unable to mount the root filesystem. Here is boot log: http://pastebin.com/jWQYHaJP

Note that card present pin can be read immediately.

Tried with two different makes of 4GB SDHC cards.

This was working last week, BTW.

If I stop boot and add rootdelay, boots OK.

setenv root /dev/mmcblk0p2 rootdelay=1
saveenv
boot

**Comment by amery** [comment]
yes, the SD driver takes around a second to react, I use `rootwait` myself.

**Comment by iainb** [comment]
I modified the mmc driver (drivers/mmc/sunxi-host/host_op.c) so that the initial timer function which detects the presence of an sd card runs immediately instead of with the default delay of 1 second. 

changing where the timer is defined within sunximmc_probe to:

smc_host->cd_timer.expires = jiffies;

This does work, I was able to boot my system without the rootwait argument passed to the kernel, though, perhaps (as this is essentially a race) I was just lucky.

However, this then causes some warnings on boot (tested on mele A1000)

[    2.990000] [mmc_pm]: no sdio card used in configuration
[    3.000000] [mmc]: sunximmc_init
[    3.010000] [mmc]: sunxi mmc controller using config : 0xb
[    3.020000] [mmc]: sunxi-mmc.0: pdev->name: sunxi-mmc, pdev->id: 0
[    3.030000] [mmc]: mmc 0 power off !!
[    3.030000] [mmc]: mmc0 Probe: base:0xdc8de000 irq:32 dma:0 pdes:0xd9a6c000, ret 0.
[    3.050000] [mmc]: mmc 0 detect change, present 1
[    3.060000] [mmc]: mmc 0 power on !!
[    3.060000] [mmc]: sunxi-mmc.1: pdev->name: sunxi-mmc, pdev->id: 1
[    3.070000] [mmc]: mmc 1 power off !!
[    3.080000] [mmc]: mmc1 Probe: base:0xdc8e0000 irq:33 dma:0 pdes:0xd9a74000, ret 0.
[    3.090000] [mmc]: sunxi-mmc.3: pdev->name: sunxi-mmc, pdev->id: 3
[    3.110000] [mmc]: mmc 3 power off !!
[    3.110000] [mmc]: sdxc_request_done(L1113): smc 0 err, cmd 52,  RTO !!
[    3.120000] [mmc]: mmc3 Probe: base:0xdc8e2000 irq:35 dma:0 pdes:0xd9a78000, ret 0.
[    3.140000] [mmc]: sdxc_request_done(L1113): smc 0 err, cmd 52,  RTO !!
[    3.160000] [mmc]: sdxc_request_done(L1113): smc 0 err, cmd 5,  RTO !!
[    3.180000] [mmc]: sdxc_request_done(L1113): smc 0 err, cmd 5,  RTO !!
[    3.200000] [mmc]: sdxc_request_done(L1113): smc 0 err, cmd 5,  RTO !!
[    3.220000] [mmc]: sdxc_request_done(L1113): smc 0 err, cmd 5,  RTO !!

Looking into drivers/mmc/sunxi-host/sdxc.c the RTO message indicates that there was a Response Timeout.

Based on the series of timeouts (possibly because we're still probing other (not used?) ports) I think that specifying the rootwait kernel argument is the right option for the moment.
